src/core/geometry/qgsgeometryutils.cpp

```diff
@@ -1257,3 +1265,533 @@ bool QgsGeometryUtils::transferFirstZValueToPoint( const QgsPointSequence &point
 
   return rc;
 }
+
+QgsPoint QgsGeometryUtils::createPointWithMatchingDimensions( double x, double y, const QgsPoint &reference )
+{
+  if ( reference.is3D() && reference.isMeasure() )
+    return QgsPoint( Qgis::WkbType::PointZM, x, y, 0.0, 0.0 );
+  else if ( reference.is3D() )
+    return QgsPoint( Qgis::WkbType::PointZ, x, y, 0.0 );
+  else if ( reference.isMeasure() )
+    return QgsPoint( Qgis::WkbType::PointM, x, y, 0.0, 0.0 );
+  else
+    return QgsPoint( x, y );
+}
+
+QgsPoint QgsGeometryUtils::interpolatePointOnSegment( double x, double y,
+    const QgsPoint &segStart, const QgsPoint &segEnd,
+    double distanceFromStart )
+{
+  QgsPoint result = createPointWithMatchingDimensions( x, y, segStart );
+
+  if ( segStart.is3D() && segEnd.is3D() && result.is3D() )
```

Comment:
This is implied by createPointWithMatchingDimensions
```suggestion
  if ( segStart.is3D() && segEnd.is3D() )
```